cpp/src/io/utilities/datasource.cpp

```diff
@@ -199,6 +200,67 @@ class direct_read_source : public file_source {
   }
 };
 
+/**
+ * @brief Implementation class for reading from a device buffer source
+ */
+class device_buffer_source final : public datasource {
+ public:
+  explicit device_buffer_source(cudf::device_span<std::byte const> d_buffer) : _d_buffer{d_buffer}
+  {
+  }
+
+  size_t host_read(size_t offset, size_t size, uint8_t* dst) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    CUDF_CUDA_TRY(cudaMemcpy(dst, _d_buffer.data() + offset, count, cudaMemcpyDeviceToHost));
```

Comment:
Does this mean we need a stream parameter in these functions?

@jrhemstad is there a downside to not using `cudaMemcpyAsync` here?

Reply:
Yup, understood. I'm trying to determine if we should add the stream parameter.

Reply:
I would vote for adding the stream parameter and using asynchronous APIs.